### PR TITLE
[feature]: TESTS, LET'S GO!

### LINKED ISSUE #14: [Feature Request]: Add actual tests
### Describe your request

We have two small, simple tests right now (both basically testing that `console.log` does what it says). Ideally we should be testing that all output is what we expect it to be (ie, the shape and content of the initial JSON data) and maybe testing the actual console output of the table. Either way, we should increase the test coverage of the whole app.

### Possible Implementation

Unit tests might be overkill for something this small (although, the structure is pretty modular -- most functionality is broken down into single-purpose functions that are exported -- so it shouldn't be annoying to implement).

For integration/E2E tests, I wonder if we could make a `sample` directory with a simple JS project (with a `package.json`, `package-lock.json` and some dependencies and actually call the CLI in that directory. In order for it to be less flaky, we could figure out how to mock the returns from the `npm` registry, but it might not be necessary as long as we lock the packages to specific versions (ie, don't use `^`, `>=` or `<=` ranges).

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kindoflew**
This PR implements adding more tests and closes #14.

# Details

- `tests` directory contains two sub directories: `integration` and `unit`
- `unit` sub directories follow `src` directory layout
- some `core` unit test directories contain `mocks` directories that themselves contain directories that have different files in different states for specific test cases (these include mock `node_modules` and mock `package.json` and `package-lock.json` files)
- the tests in directories with `mocks` directories change directory during the test (with `process.chdir`) so we can run filesystem-related tests. we need to make sure we return test runner to original `cwd` after tests
- added test documentation to `CONTRIBUTING` (including a comment about the above point on changing directories)
- ignore `log.ts` and `parse.ts` from coverage collection because those files do things that are annoying to test for and manual QA should suffice

### CODE REVIEW

File: tests/unit/cli/reporter.spec.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+const htmlExpected =
+  '<!DOCTYPEhtml><htmllang="en"><head><style>h1{font-family:arial,sans-serif;}table{font-family:arial,sans-serif;border-collapse:collapse;width:100%;}td,th{border:1pxsolid#dddddd;text-align:left;padding:8px;}tr:nth-child(even){background-color:#dddddd;}.red{color:#ff0000;}.green{color:#0f9b4e;}.yellow{color:#ce8d02;}</style><title>depngn</title></head><body><h1>Nodeversion:8.0.0</h1><table><tr><th>package</th><th>compatible</th><th>range</th></tr><tr><td>test-package-1</td><tdclass="green">true</td><td>>=8.0.0</td></tr><tr><td>test-package-2</td><tdclass="red">false</td><td>^6.0.0</td></tr><tr><td>test-package-3</td><tdclass="yellow">undefined</td><td>n/a</td></tr></table></body></html>';
+
+describe('createReport', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('outputs correct table', async () => {
+    const originalConsoleLog = console.log;
```
Comment (KostiantynPopovych) [inline tests/unit/cli/reporter.spec.ts]: May we move this logic to the `beforeEach` and `afterEach` functions?

File: tests/unit/cli/reporter.spec.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import { createReport } from '../../../src/cli/reporter';
```
Comment (KostiantynPopovych) [inline tests/unit/cli/reporter.spec.ts]: Any chance to migrate to the usage of aliases/absolute paths?

File: tests/unit/cli/reporter.spec.ts (hunk 1 above)
Comment (kindoflew) [reply tests/unit/cli/reporter.spec.ts]: oh yeah good call!

usually, i like to keep these in each individual test -- to keep things explicit and obvious -- but for these tests, we'll always have to mock `console.log`, so i think it makes sense.

File: tests/unit/cli/reporter.spec.ts (hunk 2 above)
Comment (kindoflew) [reply tests/unit/cli/reporter.spec.ts]: yeah, i think it's time lol

i can add it in this PR

**Pull request merged**